Decouple property naming from PropertyCollection

Spea opened the pull request:
While working on discriminator support, I stumbled upon a bug where it was impossible to force a `camelCase` name for a property.

So using a class like this:

```php
class MyClass {
    #[SerializedName('myProperty')]
    public string $myProperty;
}
```

would result in the following error

```
LogicException : You can not rename my_property into myProperty as it is the same property. Did you miss to handle camelCase properties with PropertyCollection::serializedName?

```

With the changes I did in this PR, it is now possible to pass different property naming strategies to the different parsers.
Besides fixing the bug at hand, this also provides more freedom when it comes to property collection handling, as you don't have to be aware that the collection itself creates the serialized name, but instead it is up to the caller to decide what serialized name should be used.

Spea committed:
Add test for using camelCase serialized name

There is currently a bug where when you use a camelCase name for the `SerializedName` attribute which would result in the same snake_case name, that the `RawClassMetadata::renameProperty` method would throw an exception because it thinks we are trying to use the same name again.
This test only acts as a reference so we can fix the actual bug.

Spea committed:
Decouple property naming from `PropertyCollection`

With this change it is now possible to pass different property naming strategies to the different parsers.
This also provides more freedom when it comes to property collection handling, as you don't have to be aware that the collection itself creates the serialized name, but instead it is up to the caller to decide what serialized name should be used.

dbu commented on src/ModelParser/PhpDocParser.php:
```diff
@@ -55,8 +61,9 @@ private function parseProperties(\ReflectionClass $reflClass, RawClassMetadata $
+                $serializedName = $this->namingStrategy->getSerializedName($reflProperty->getName());
                 $property = PropertyVariationMetadata::fromReflection($reflProperty);
-                $classMetadata->addPropertyVariation($reflProperty->getName(), $property);
+                $classMetadata->addPropertyVariation($serializedName, $property);
```
oh, i vaguely remember that we struggled with different representation of the same property.

so your approach is to always use the serialized name in the metadata and never the original name?

Spea replied:
> so your approach is to always use the serialized name in the metadata and never the original name?

That was actually already the case, the only difference now is, that the serialized name is created directly in the parser so you have more control over it instead of directly in the [PropertyCollection](https://github.com/liip/metadata-parser/blob/2.x/src/ModelParser/RawMetadata/PropertyCollection.php#L23)

dbu reviewed (commented):
thanks for this pull request. seems like a nice cleanup, that static method was indeed rather ugly.

i am slightly worried that when different parsers are given different strategies (e.g. one is forgotten and the default is taken), the parsers would add metadata to different names, leading to confusion. but i don't see a good way around that, i think we just need to document that.

can you please adjust the readme section https://github.com/liip/metadata-parser?tab=readme-ov-file#property-naming-strategy (haha, were we call it strategy when its a global static flag :-P - you actually convert that to the strategy pattern here)

Spea commented:
> i am slightly worried that when different parsers are given different strategies (e.g. one is forgotten and the default is taken), the parsers would add metadata to different names, leading to confusion

I thought about this too, and I see a few other directions we could go here

* Create an abstract parser class which handles the default strategy and has a separate method for injecting a different one
* Create a separate interface `ModelParserNamingStrategyAwareInterface` which has one method `setPropertyNamingStrategy` and check in the `Builder` class if the current parser implements this interface and call the related method
* Add a second parameter to the `parse` method where you have to pass the naming strategy
  ```php
  parse(RawClassMetadata $classMetadata, PropertyNamingStrategyInterface $namingStrategy): void
   ```

This means in all cases the `Builder` class would be the responsible one for handling naming strategies.

Personally I think the third option is better, but also a bit weird in case the parser does not need any naming strategy (like the [LiipMetadataAttributeParser](https://github.com/liip/metadata-parser/blob/b2897803104cc399b317c0a685d586d14e10bd78/src/ModelParser/LiipMetadataAttributeParser.php))

What do you think?

dbu commented:
i like the third option best. yes, we will pass the strategy even when no naming is needed, but there is no real cost involved as the strategy is injected/created only once. 
the consistency of the strategy is very important, so being able to guarantee that seems worth it.

dbu commented:
do you adjust this PR correspondingly? 

are there other things for version 2.x or should i tag the release once we have these changes?

Spea commented:
> do you adjust this PR correspondingly?
> 
> are there other things for version 2.x or should i tag the release once we have these changes?

Yes, I will adjust this PR accordingly. After this, the last thing I want to add is the discriminator support (which is almost done, at least the metadata-parser side). However, I have an upcoming vacation, so I'm not sure if I can finish it before, but I will definitely finish it this month.

Spea committed:
Pass naming strategy to parse method

This ensures that the same strategy is used for all parsers. Using the strategy is completely optional, so parsers can just ignore it if they don't need it.

Spea commented:
> do you adjust this PR correspondingly?

Done via 2006b3d168ff4b87b660ceacaa1df568a1c217d6 - Thanks for your feedback on this 🙇

dbu commented on CHANGELOG.md:
```diff
@@ -12,6 +12,11 @@
 * Adjusted code to not trigger warnings with PHP 8.4
 * Removed deprecated `getDeserializeFormat` method from date time
 * Replaced `@Preferred` annotation with the `#[Preferred]` attribute
+* Added a new `PropertyNamingStrategyInterface`  which can be used to modify the naming strategy for properties. Previously 
```
```suggestion
* Replaced `PropertyCollection::useIdenticalNamingStrategy` with the `PropertyNamingStrategyInterface`. To change the default strategy, you can now inject it into the parser instead of changing a static property.
```

dbu commented on CHANGELOG.md:
```diff
@@ -12,6 +12,11 @@
 * Removed deprecated `getDeserializeFormat` method from date time
 * Replaced `@Preferred` annotation with the `#[Preferred]` attribute
+* Added a new `PropertyNamingStrategyInterface`  which can be used to modify the naming strategy for properties. Previously 
+  this could be adjusted via the `PropertyCollection::useIdenticalNamingStrategy` method, but as this was a bit error prone, a separate
```
```suggestion
```

dbu commented on CHANGELOG.md:
```diff
@@ -12,6 +12,11 @@
 * Replaced `@Preferred` annotation with the `#[Preferred]` attribute
+* Added a new `PropertyNamingStrategyInterface`  which can be used to modify the naming strategy for properties. Previously 
+  this could be adjusted via the `PropertyCollection::useIdenticalNamingStrategy` method, but as this was a bit error prone, a separate
+  interface was introduced. This new interface also comes with two implementations: 
```
```suggestion
  The library provides two implementations for the property naming strategy: 
```

dbu commented on CHANGELOG.md:
```diff
@@ -12,6 +12,11 @@
+  this could be adjusted via the `PropertyCollection::useIdenticalNamingStrategy` method, but as this was a bit error prone, a separate
+  interface was introduced. This new interface also comes with two implementations: 
+  * `IdenticalPropertyNamingStrategy` 
+  * `SnakeCasePropertyNamingStrategy`.
```
```suggestion
  * `SnakeCasePropertyNamingStrategy` (default).
```

dbu reviewed (approved):
great, thanks a lot! i did some tweaks with the changelog, will try to accept those changes myself and merge.

i am in no hurry to release version 2, i will wait for the discriminator support. it will be easier if its in a major version, and maybe we end up wanting to do BC breaks.

dbu commented:
i applied the changelog changes and merged the branch. thanks a lot for this contribution !
